primitives/src/addr.rs

```diff
@@ -100,12 +190,586 @@ impl core::ops::Index<usize> for Ipv6Addr {
     }
 }
 
+impl core::str::FromStr for Ipv6Addr {
+    type Err = ParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
```

Comment:
While this is acceptable, The trait is implemented for `std::net::SocketAddrv6` (and v4). You also know that this is only used with `serde` with the `std` feature enabled. As such you can avoid having to do all this work, by deferring to those primitives under a feature flag.

Reply:
Good catch, will do

Reply:
Actually, deserialization creeps into `iroha_data_model::Peer`, which would also then need to implement Deserialize conditionally on `std`. That seems like an inconsistent API to me.

Reply:
IMO, the `Peer` as a whole is an oversight if allowed on the smartcontract side: reason being that the `Peer` operations are registering and de-registering, which **should** be done rarely and with human oversight. 

Moreover, any form of `serde` deserialization has to do with running code in the peer, (not with the smartcontract) which usually depends on `std`. I don't think it's inconsistent, it's just not exactly what we're doing at the moment. 

If you'd like this to be a separate issue, I agree, but not that it shouldn't be done in general.

Reply:
> IMO, the Peer as a whole is an oversight if allowed on the smartcontract side: reason being that the Peer operations are registering and de-registering, which should be done rarely and with human oversight.

Runtime validators will want to use it though, and they share API surface with smartcontracts AIUI.

> Moreover, any form of serde deserialization has to do with running code in the peer, (not with the smartcontract) which usually depends on std. I don't think it's inconsistent, it's just not exactly what we're doing at the moment.

That I agree with, probably losing serde support for `Peer` in `no_std` environment is not a big deal. I'll update the PR.

Reply:
> If you'd like this to be a separate issue, I agree, but not that it shouldn't be done in general.

On further examination, I'd prefer to make this a separate issue. It chains into a lot of changes in `data_model`. Opened https://github.com/hyperledger/iroha/issues/3407